library/std/src/io/stdio.rs

```diff
@@ -1017,6 +1018,145 @@ where
     }
 }
 
+mod terminal_sealed {
+    #[unstable(feature = "sealed", issue = "none")]
+    pub trait Sealed {}
+}
+
+/// Trait to determine if a descriptor/handle refers to a terminal/tty.
+#[unstable(feature = "is_terminal", issue = "98070")]
+pub trait IsTerminal: terminal_sealed::Sealed {
+    /// Returns true if the descriptor/handle refers to a terminal/tty.
+    ///
+    /// Returns an error if any errors occurred when attempting to make this determination, other
+    /// than those expected as part of making this determination. For instance, on some platforms
+    /// checking for a terminal may involve making a system call that returns an error on a
+    /// non-terminal, so that error will get translated to `Ok(false)`, but any unexpected error
+    /// will still get returned as an error.
+    ///
+    /// On platforms where Rust does not know how to detect a terminal yet, this will return
+    /// `Ok(false)`.
+    fn is_terminal(&self) -> io::Result<bool>;
+}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for File {}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for Stdin {}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for StdinLock<'_> {}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for Stdout {}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for StdoutLock<'_> {}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for Stderr {}
+
+#[unstable(feature = "sealed", issue = "none")]
+impl terminal_sealed::Sealed for StderrLock<'_> {}
+
+cfg_if::cfg_if! {
+    if #[cfg(any(unix, target_os = "wasi"))] {
+        use crate::os::fd::owned::{AsFd, BorrowedFd, OwnedFd};
+        use crate::os::fd::raw::AsRawFd;
+
+        #[unstable(feature = "sealed", issue = "none")]
+        impl terminal_sealed::Sealed for BorrowedFd<'_> {}
+
+        #[unstable(feature = "sealed", issue = "none")]
+        impl terminal_sealed::Sealed for OwnedFd {}
+
+        #[unstable(feature = "is_terminal", issue = "98070")]
+        impl IsTerminal for BorrowedFd<'_> {
+            fn is_terminal(&self) -> io::Result<bool> {
+                let ret = unsafe { libc::isatty(self.as_raw_fd()) };
+                if ret == 0 {
+                    let err = io::Error::last_os_error();
+                    if err.raw_os_error() == Some(libc::ENOTTY) {
```

Comment:
The Linux documentation [here](https://man7.org/linux/man-pages/man3/isatty.3.html#ERRORS) says that "older kernels" return `EINVAL` here. Do you happen to know which versions those are and if they're supported by Rust?